### PR TITLE
Test annotation for endpoints

### PR DESCRIPTION
**Pull request by lfportal**
Added the `@test` annotation for specifying an interaction for testing purposes. This includes a parser and cleanser components. Verification rules for the tests are not enforced in this PR.

### CODE REVIEW

File: lib/src/models/definitions.ts [hunk 1]
```diff
@@ -72,3 +73,34 @@ export interface BodyDefinition {
   description?: string;
   type: DataType;
 }
+
+export interface TestDefinition {
+  description?: string;
+  states?: TestStateDefinition[];
```
Comment (fwouts) [inline lib/src/models/definitions.ts]: Shouldn't this be required, but with an empty list when absent?

File: lib/src/models/definitions.ts [hunk 2]
```diff
@@ -72,3 +73,34 @@ export interface BodyDefinition {
...
+  request?: TestRequestDefinition;
+  response: TestResponseDefinition;
+}
+
+export interface TestStateDefinition {
+  name: string;
+  params?: { name: string; expression: DataExpression }[];
```
Comment (fwouts) [inline lib/src/models/definitions.ts]: ditto (also each one below)

File: lib/src/models/nodes.ts [hunk 3]
```diff
@@ -73,3 +74,34 @@ export interface BodyNode {
   description?: Locatable<string>;
   type: DataType;
 }
+
+export interface TestNode {
+  description?: Locatable<string>;
+  states?: TestStateNode[];
```
Comment (fwouts) [inline lib/src/models/nodes.ts:87]: ditto

File: lib/src/parsers/utilities/expression-parser.spec.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            }
+          ]
+        },
+        { kind: "null" }
+      ],
+      kind: "array"
+    });
+  });
+
+  test("parses a object literal expression", () => {
```
Comment (fwouts) [inline lib/src/parsers/utilities/expression-parser.spec.ts]:
```suggestion
  test("parses an object literal expression", () => {
```

File: lib/src/parsers/utilities/parser-utility.ts [hunk 5]
```diff
@@ -173,6 +186,83 @@ export function extractNumberProperty(
...
+    }
+    const value = property.getInitializerIfKindOrThrow(
+      ts.SyntaxKind.ArrayLiteralExpression
+    );
+    const location = property.getSourceFile().getFilePath();
+    const line = value.getStartLineNumber();
+
+    return { value, location, line };
+  } else {
+    return;
```
Comment (fwouts) [inline lib/src/parsers/utilities/parser-utility.ts]: Maybe `return undefined` for clarity? You could also just remove this entirely (`return undefined` is the default behaviour).

**Review by fwouts — CHANGES_REQUESTED** [review]
Nice work. Would be good to tweak it to add type safety later :)

File: lib/src/models/nodes.ts (hunk 3 above)
Comment (lfportal) [reply lib/src/models/nodes.ts:87]: Planning to introduce `Locatable` to some of the `@test` related nodes in the future

File: lib/src/models/nodes.ts [hunk 6]
```diff
@@ -80,3 +81,34 @@ export interface BodyNode {
...
+  request?: Locatable<TestRequestNode>;
+  response: Locatable<TestResponseNode>;
+}
+
+export interface TestStateNode {
+  name: string;
+  params?: { name: string; expression: DataExpression }[];
```
Comment (fwouts) [inline lib/src/models/nodes.ts:94]: Should these (and below in TestRequestNode and TestResponseNode) also not be optional?

**Review by fwouts — APPROVED** [review]

**Pull request merged**